docs/development/design/build-images.rst

```diff
@@ -48,131 +45,242 @@ This produced issues to people pinning their images to any of these names becaus
-  * other languages version managers (``pyenv``, ``nodenv``, etc)
+  * languages version manager (``asdf``) and its plugins for each language
   * UID and GID
 
-The following images all are based on ``ubuntu20-base``:
+Instead of building all the Docker image per language versions combination,
+it will be easier to install all of them at build time using the same steps.
+Installing a language only adds a few seconds when binaries are provided.
+However, to reduce the time to install these languages as much as possible,
+a local mirror hosted on S3 for each language will be created.
 
-* ``ubuntu20-py*``
+It's important to note that Python does not provide binaries and compiling a version takes around ~2 minutes.
+However, the Python versions could be pre-compiled and expose their binaries via S3 to builders.
+Then, at build time, the builder will only download the binary and copy it in the correct path.
 
-  * Python version installed via ``pyenv``
-  * default Python packages (pinned versions)
-    * pip
-    * setuptools
-    * virtualenv
-  * labels
-
-* ``ubuntu20-conda*``
+.. note::
 
-  * same as ``-py*`` versions
-  * Conda version installed via ``pyenv``
-  * ``mamba`` executable (installed via ``conda``)
+   Depending on the demand, Read the Docs may pre-build the most common combinations of languages used by users.
+   For example, ``ubuntu20-python39+node14`` or ``ubuntu20-python39+node14+rust1``.
+   However, this is seen as an optimization for the future and it's not required for this document.
 
-Note that all these images only need to run ``pyenv install ${PYTHON_VERSION}``
-to install a specific Python/Conda version.
 
-.. Build all these images with Docker
+Build steps
+-----------
 
-   docker build -t readthedocs/build:ubuntu20-base -f Dockerfile.base .
-   docker build -t readthedocs/build:ubuntu20-py39 -f Dockerfile.py39 .
-   docker build -t readthedocs/build:ubuntu20-conda47 -f Dockerfile.conda47 .
+With this new approach, the steps followed by a builder will be:
 
-   Check the shared space between images
-   docker system df --verbose | grep -E 'SHARED SIZE|readthedocs'
+#. run some application code to spin up the ``-base`` Docker image into a container
+#. execute ``git`` inside the container to clone the repository
+#. analyze and parse files (``.readthedocs.yaml``) from the repository *outside* the container
+#. spin up a new Docker container based on the Ubuntu OS specified in the config file
+#. install all language dependencies from the cache
+#. create the environment and install docs' dependencies inside the container
+#. execute build commands inside the container
+#. push the output generated by build commands to the storage
 
-   Initial Dockerfile.* as example for this are pushed in this PR
-   https://github.com/readthedocs/readthedocs-docker-images/pull/166
+The main difference with the current approach are:
 
+* the image spin is selected depending on the OS version
```

Comment:
Typo here, not sure what `spin` is meant to be.

Reply:
"spin up an instance" or "spin up a container with an image". I guess that just "spin" doesn't work here, tho 😄